sdks/python/apache_beam/runners/portability/stager.py

```diff
@@ -0,0 +1,573 @@
+
+  def _copy_file(self, from_path, to_path):
+    """Copies a local file to a GCS file or vice versa."""
+    logging.info('file copy from %s to %s.', from_path, to_path)
+    if from_path.startswith('gs://') or to_path.startswith('gs://'):
+      from apache_beam.io.gcp import gcsio
+      if from_path.startswith('gs://') and to_path.startswith('gs://'):
+        # Both files are GCS files so copy.
+        gcsio.GcsIO().copy(from_path, to_path)
+      elif to_path.startswith('gs://'):
+        # Only target is a GCS file, read local file and upload.
+        with open(from_path, 'rb') as f:
+          with gcsio.GcsIO().open(to_path, mode='wb') as g:
+            pfun = functools.partial(f.read, gcsio.WRITE_CHUNK_SIZE)
+            for chunk in iter(pfun, ''):
+              g.write(chunk)
+      else:
+        # Source is a GCS file but target is local file.
+        with gcsio.GcsIO().open(from_path, mode='rb') as g:
+          with open(to_path, 'wb') as f:
+            pfun = functools.partial(g.read, gcsio.DEFAULT_READ_BUFFER_SIZE)
+            for chunk in iter(pfun, ''):
+              f.write(chunk)
+    else:
+      # Branch used only for unit tests and integration tests.
+      # In such environments GCS support is not available.
+      if not os.path.isdir(os.path.dirname(to_path)):
+        logging.info(
+            'Created folder (since we have not done yet, and any errors '
+            'will follow): %s ', os.path.dirname(to_path))
+        os.mkdir(os.path.dirname(to_path))
+      shutil.copyfile(from_path, to_path)
+
+  def _download_file(self, from_url, to_path):
+    """Downloads a file over http/https from a url or copy it from a remote
+        path to local path."""
+    if from_url.startswith('http://') or from_url.startswith('https://'):
+      # TODO(silviuc): We should cache downloads so we do not do it for every
+      # job.
+      try:
+        # We check if the file is actually there because wget returns a file
+        # even for a 404 response (file will contain the contents of the 404
+        # response).
+        # TODO(angoenka): Extract and use the filename when downloading file.
+        response, content = __import__('httplib2').Http().request(from_url)
+        if int(response['status']) >= 400:
+          raise RuntimeError(
+              'Artifact not found at %s (response: %s)' % (from_url, response))
+        with open(to_path, 'w') as f:
+          f.write(content)
+      except Exception:
+        logging.info('Failed to download Artifact from %s', from_url)
+        raise
+    else:
+      # Copy the file from the remote file system to loca files system.
+      self._copy_file(from_url, to_path)
+
+  def _is_remote_path(self, path):
+    return path.find('://') != -1
+
+  def _stage_extra_packages(self, extra_packages, staging_location, temp_dir):
+    """Stages a list of local extra packages.
+
+      Args:
+        extra_packages: Ordered list of local paths to extra packages to be
```

Comment:
<!--new_thread; commit:fb22e20e257e1c00a618286cb0a1dd01e4071709; resolved:0-->
Please clarify that currently supported source location are GCS, and local files. HTTP locations for extra_packages are currently not supported.